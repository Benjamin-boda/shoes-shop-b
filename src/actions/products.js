import database from "../firebase/firebase";
import firebase from 'firebase/app'

//ADD_PRODUCT

export const addProduct = (product) => ({
    type: "ADD_PRODUCT",
    product
});

export const startAddProduct = (productData) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        const {
            image = "",
            name = "",
            size = "",
            price = 0,
            boughtAt = 0
        } = productData;
        const product = {image, name, size, price, boughtAt};

        return database.ref(`users/${uid}/products`).push(product).then((ref) => {
            dispatch(addProduct({
                id: ref.key,
                ...product
            }))
        })
    }
}

//REMOVE_PRODUCT

export const removeProduct = ({id} = {}) => ({
    type: "REMOVE_PRODUCT",
    id
});

export const startRemoveProduct = ({id}) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        return database.ref(`users/${uid}/products/${id}`).remove().then(() => {
            dispatch(removeProduct({id}))
        })
    }
}

//EDIT_PRODUCT

export const editProduct = (id, updates) => ({
    type: "EDIT_PRODUCT",
    id,
    updates
});

export const startEditProduct = (id, updates) => {
    return (dispatch) => {
        const uid = firebase.auth().currentUser.uid;
        return database.ref(`users/${uid}/products/${id}`).update(updates).then(() => {
            dispatch(editProduct(id, updates))
        })
    }
}

//SET_PRODUCTS

export const setProducts = (products) => ({
    type: "SET_PRODUCTS",
    products
})

export const startSetProducts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/products`).once("value").then((snapshot) => {
            const products = [];

            snapshot.forEach((childSnapshot) => {
                products.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
        dispatch(setProducts(products))
        })
    }
}

//REMOVE ALL

export const resetProducts = () => ({
    type: "RESET_PRODUCTS"
});

export const startResetProducts = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        return database.ref(`users/${uid}/products`).remove()
            .then(() => {
                dispatch(resetProducts());
            });
    };
};