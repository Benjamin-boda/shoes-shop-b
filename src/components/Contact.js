import React, { useEffect, useState } from "react";

export const Contact = () => {
    const [subject, setSubject] = useState("")
    const [description, setDescription] = useState("")
    const [submit, setSubmit] = useState("")

    const inputChange = (e) => {
        setSubject(e.target.value)
    }

    const textChange = (e) => {
        setDescription(e.target.value)
    }

    useEffect(() => {
        const request = {
            title: subject,
            object: description
        }
        setSubmit(request)
    }, [subject, description])

    const onSubmit = () => {
        console.log(submit)
    }

    return (
        <div className="contact">
            <div className="contact__header">
                <h1 className="contact__title">Contact us</h1>
                <img className="contact__image" src={require("../images/conctact logo.jpg").default}/>
            </div>
            
            <form className="contact__form" onSubmit={onSubmit}>
                <input data-testid="input" className="contact__form__subject" onChange={inputChange} type="text" placeholder="Subject"/>
                <textarea data-testid="textarea" className="contact__form__description" onChange={textChange} maxLength="200" minLength="10" placeholder="Describe your problem in 200 caracters"/>
                <button className="contact__form__button">Submit</button>
            </form>
        </div>
    )
}