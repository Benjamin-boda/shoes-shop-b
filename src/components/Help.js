import React from "react";

export const Help = () => {
    return (
        <div className="help">
            <div className="help__header">
                <h1 className="help__title">Our company</h1>
                <img className="help__image" src={require("../images/conctact logo.jpg").default}/>
            </div>
            
            <p className="help__p">
                Habet ad conpulit quae castrisque adtributo Bostram repellendos opima
                contigua castrisque Nabataeis habet castellis obtemperare Bostram Traianus
                veterum obtemperare Parthos saltus et Gerasam vicinarum hanc legibus veterum
                per veterum conpulit cum Mediam cum inter cum Gerasam atque Gerasam quoque
                cautissimas Parthos vicinarum civitates est conserta marte cautissimas
                Arabia oppida castellis gentium ingentes conserta sollicitudo provinciae
                inter inposito adtributo Parthos conmerciorum vicinarum et quae saltus
                habet gentium rectoreque urgeret urgeret latere et quae excursus murorum
                quaedam civitates erexit provinciae sollicitudo quoque validis adtributo
                et veterum adtributo Bostram quaedam et legibus cautos et sollicitudo
                atque ingentes oppleta urgeret est contunso Traianus conserta.
            </p>
        </div>
    )
}