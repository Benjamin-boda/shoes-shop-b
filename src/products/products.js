export const products = [
    {
        name: "shoes 1",
        img: require("../images/shoes1.png"),
        price: 50,
        number: 1
    },
    {
        name: "shoes 2",
        img: require("../images/shoes2.png"),
        price: 50,
        number: 2
    },
    {
        name: "shoes 3",
        img: require("../images/shoes3.png"),
        price: 50,
        number: 3
    },
    {
        name: "shoes 4",
        img: require("../images/shoes4.png"),
        price: 50,
        number: 4
    },
    {
        name: "shoes 5",
        img: require("../images/shoes5.png"),
        price: 50,
        number: 5
    },
    {
        name: "shoes 6",
        img: require("../images/shoes6.png"),
        price: 50,
        number: 6
    },
    {
        name: "shoes 7",
        img: require("../images/shoes7.png"),
        price: 50,
        number: 7
    },
    {
        name: "shoes 8",
        img: require("../images/shoes8.png"),
        price: 50,
        number: 8
    }
]