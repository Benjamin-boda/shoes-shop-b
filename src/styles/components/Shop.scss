.shop {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    margin: 0 auto;
}

.shop__shoes {
    align-items: center;
    display: flex;
    flex-direction: column;
    height: 20vh;
    justify-content: center;
    margin: 0;
    text-decoration: none;
    width: 100%;

    @media (min-width: $desktop-breakpoint-large) {
        height: initial;
        width: 49.3vw;
    }
}

.shop__shoes__0 {
    background: $cream;
}

.shop__shoes__0:hover {
    background: darken($cream, 5%);
}

.shop__shoes__1 {
    background: $cashmere;
}

.shop__shoes__1:hover {
    background: darken($cashmere, 5%);
}

.shop__shoes__2 {
    background: $flora;
}

.shop__shoes__2:hover {
    background: darken($flora, 5%);
}

.shop__shoes__3 {
    background: $ivory;
}

.shop__shoes__3:hover {
    background: darken($ivory, 5%);
}

.shop__image {
    height: auto;
    transition: transform .3s;
    width: 25vw;
}

.shop__image:hover {
    transform: scale(1.2);
}

.shop__name {
    color: white;
    font-family: "MonserratRegular";
    font-size: $font-size-xsmall;
}

.shop__pages {
    background:lighten(#32485C, 20%);
    height: 5vh;
    text-align: center;

    @media (min-width: $desktop-breakpoint-large) {
        height: initial;
    }
}

.shop__pages__link {
    color: white;
    font-family: "MonserratBlack";
    font-size: $font-size-small;
    margin: 0 $s-size;
    text-decoration: none;
}

.shop__pages__link:hover {
    color: black;
}