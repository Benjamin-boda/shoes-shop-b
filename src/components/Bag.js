import React, { useEffect, useState } from "react";
import {useSelector, useDispatch} from "react-redux";
import {startRemoveProduct, startEditProduct, startResetProducts} from "../actions/products";
import {startAddOrder} from "../actions/orders";
import products from "../tests/fixtures/products";

export const Bag = () => {
    const [sizeUpdate, setSizeUpdate] = useState("")
    const [productUpdate, setProductUpdate] = useState("")
    const productsInStore  = useSelector(state => state.products)
    const dispatch = useDispatch()

    const sizes = [37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47]
    

    const sizeEdit = (e) => {
        setSizeUpdate(e.target.value)
    }

    useEffect(() => {
        if (productUpdate !== "" & productsInStore.lenght > 0) {
            const filterProductArray = productsInStore.filter((product) => productUpdate === product.name)
            const productEdit = {
                image : {
                    default : filterProductArray[0].image.default
                },
                name : filterProductArray[0].name,
                size : sizeUpdate,
                price : filterProductArray[0].price,
                boughtAt : filterProductArray[0].boughtAt
            }
            const id = filterProductArray[0].id
            dispatch(startEditProduct(id, productEdit))    
        }
    }, [sizeUpdate])

    const paymentOrder = () => {
        productsInStore.map((product) => {
            product.boughtAt = Date()
            dispatch(startAddOrder(product))})
        dispatch(startResetProducts())              
    }

    const sumPrice = productsInStore.length > 0 ? productsInStore.map(product => product.price).reduce((acc, price) => acc + price) : "0"

    return (
        <div className="bag__content">
            <div className="bag__box">
                {productsInStore.length > 0 ?
                    productsInStore.map((product) => 
                        <div className="bag" key={product.name} onMouseEnter={() => setProductUpdate(product.name)}>
                            <img className="bag__image" src={product.image.default}/>
                            <p data-testid={product.name} className="bag__name">{product.name}</p>
                            <select className="bag__select" name="size" onChange={sizeEdit}>
                                <option value={product.size}>{product.size}</option>
                                {sizes.map((size) => 
                                    <option value={size} key={size}>{size}</option>
                                    )}
                            </select>
                            <p className="bag__name">Price : {product.price} €</p>
                            <button className="bag__button" onClick={() => dispatch(startRemoveProduct(product))}><img className="bag__cross__button" src={require("../images/cross button.png").default}/></button>
                        </div>
                    ) : <p className="bag__empty">Nothing in the bag</p>
                }
            </div>
            
            <p className="bag__price">Total price : {sumPrice} €</p>

            <div className="bag__payment__div">
                <button className="bag__payment" onClick={paymentOrder} disabled={productsInStore.length === 0 ? true : false}>Payment</button>
            </div>
            
        </div>
    )
}