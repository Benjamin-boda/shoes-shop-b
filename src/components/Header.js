import React, {useState ,useContext} from "react";
import {Link} from "react-router-dom";
import {useDispatch, useSelector} from "react-redux";
import {startLogin, startLogout} from "../actions/auth";
import {AuthContext} from "../firebase/Auth";
import {LoginModal} from "./LoginModal";

export const Header = () => {
    const {currentUser} = useContext(AuthContext);
    const [accountHovering, setAccountHovering] = useState(false)
    const [modalIsOpen, setModalIsOpen] = useState(false)
    const productsInStore = useSelector(state => state.products)
    const dispatch = useDispatch()

    const openModal = () => {
        if (!!currentUser) {
            setModalIsOpen(false)
        } else {
            setModalIsOpen(true)
        }
    }

    return (
        <div className="header">
            <img className="header__logo" src={require("../images/logo.png").default}/>
            <Link className="header__link" to="/">Home</Link>
            <Link className="header__link" to={!!currentUser ? "/shop/1" : ""} onClick={openModal}>Shop</Link>
            <Link className="header__link" to="/contact">Contact</Link>
            <Link className="header__link" to="/help">Help</Link>
            
            <div className="header__account">
                <Link 
                    data-testid="account"
                    className="header__link__account"
                    to=""
                    onClick={() => !!currentUser ? undefined : dispatch(startLogin())}
                    onMouseEnter={() => !!currentUser ? setAccountHovering(true) : setAccountHovering(false)}
                >
                    {!!currentUser ? "User account" : "Login"}
                </Link>
                {accountHovering ? 
                    <ul className="header__dropdown" onClick={() => setAccountHovering(false)} onMouseLeave={() => setAccountHovering(false)}>
                        <Link className="header__link__account__dropdown" to="/orders">
                            <li className="header__dropdown__item">My account</li>
                        </Link>
                        <li className="header__dropdown__item" onClick={() => dispatch(startLogout())}>Logout</li>
                        <li className="header__dropdown__item" onClick={() => setAccountHovering(false)}>Close</li>
                    </ul> : undefined
                }
            </div>
            
            <div>
                <Link to={!!currentUser ? "/bag" : ""} onClick={openModal}><img className="header__bag" src={require("../images/bag.png").default}/></Link>
                <p className="header__bag__counter">{productsInStore.length > 0 ? productsInStore.length : 0}</p>
            </div>
            

            <LoginModal modalIsOpen={modalIsOpen} setModalIsOpen={setModalIsOpen}/>
        </div>
    )
}