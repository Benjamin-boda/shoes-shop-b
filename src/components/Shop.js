import React, {useState} from "react";
import {Link} from "react-router-dom";
import {products} from "../products/products";

export const Shop = () => {
    const [currentPage, setCurrentPage] = useState(1)
    const numberPage = Math.ceil(products.length/4)
    let pages = []

    for (let i = 1; i < numberPage + 1; i++) {
        pages.push(i)
    }

    return (
        <div>
            <div className="shop">
                {products.slice((currentPage - 1) * 4, currentPage * 4)
                    .map((product, index) =>
                    <Link className={`shop__shoes shop__shoes__${index}`} to={`/product/${product.number}`} key={product.number}>
                        <img className="shop__image" src={product.img.default}/>
                        <p className="shop__name">{product.name}</p>
                    </Link>
                )}
            </div>
            
            <div className="shop__pages">
                {pages.map((page) => 
                    <Link className="shop__pages__link" key={page}  onClick={() => setCurrentPage(page)} to={`/shop/${page}`}>
                        {page}
                    </Link>
                    )}
            </div>
            
        </div> 
    )
}