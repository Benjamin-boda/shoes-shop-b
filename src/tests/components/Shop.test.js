import React from "react";
import {Shop} from "../../components/Shop";
import {render, cleanup, screen, fireEvent} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect';
import products from "../fixtures/products";
import { MemoryRouter } from 'react-router-dom';

afterEach(cleanup);

const { asFragment } = render(
    <MemoryRouter>
        <Shop products={products}/>
    </MemoryRouter>)

test("should match the snapshot", () => {
    
    expect(asFragment(<Shop/>)).toMatchSnapshot()
})

test("should contain this text", () => {
    const { getByText } = render(
    <MemoryRouter>
        <Shop products={products}/>
    </MemoryRouter>)

    expect(getByText("shoes 1")).toBeTruthy()
})