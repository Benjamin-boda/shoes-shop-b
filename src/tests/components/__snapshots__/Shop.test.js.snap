// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="shop"
    >
      <a
        class="shop__shoes shop__shoes__0"
        href="/product/1"
      >
        <img
          class="shop__image"
        />
        <p
          class="shop__name"
        >
          shoes 1
        </p>
      </a>
      <a
        class="shop__shoes shop__shoes__1"
        href="/product/2"
      >
        <img
          class="shop__image"
        />
        <p
          class="shop__name"
        >
          shoes 2
        </p>
      </a>
      <a
        class="shop__shoes shop__shoes__2"
        href="/product/3"
      >
        <img
          class="shop__image"
        />
        <p
          class="shop__name"
        >
          shoes 3
        </p>
      </a>
      <a
        class="shop__shoes shop__shoes__3"
        href="/product/4"
      >
        <img
          class="shop__image"
        />
        <p
          class="shop__name"
        >
          shoes 4
        </p>
      </a>
    </div>
    <div
      class="shop__pages"
    >
      <a
        class="shop__pages__link"
        href="/shop/1"
      >
        1
      </a>
      <a
        class="shop__pages__link"
        href="/shop/2"
      >
        2
      </a>
    </div>
  </div>
</DocumentFragment>
`;
